### ISSUE TITLE
Disabling branch coverage on a subset of tests

### ISSUE THREAD

**Issue by opikalo** [gcovr/gcovr#482]
**Current problem**
Right now we are using gcovr with SonarQube to collect coverage on C++ files. Sonarqube offers three options for test code: 
1) analyze tests the same way as sources
2) ignore tests entirely from analysis (and ignore test coverage)
3) Use 1) and ignore reporting coverage from tests 

We are using option 1) because it provides a comprehensive insight on which tests. But because SonarQube is using branch coverage, on test metrics it becomes very confusing: statements like `assert(True)`  will be reported as missing coverage.

**Describe the solution you'd like**
I'd like to specify a pattern for files that will drop branch coverage from the report: and report line coverage only. When generating a report, a file under consideration will be tried against the pattern, and if there is a match, it will be treated as "no-branch" line only reporting. 

**Describe alternatives you've considered**
One possibility is to scrub branch information from the XML report after the generation is done. Another possibility is to run gcovr on some files with branch options, and without branch option on test files, and combine the two (or specify two sources to sonarqube for processing. There is some overhead either way. 

**Additional context**
I'm interested in implementing this feature, but I want to solicit a round of feedback whether gcovr is the right place to be doing this.

**Comment by opikalo** [comment]
Looks like there is somewhat similar #324, but here I'm interested in specifying a pattern of files (C++ tests) for which there should be no branch coverage reported.

**Comment by latk** [comment]
TL;DR: workarounds are available. Adding such functionality to gcovr in a good way seems difficult.

Currently, gcovr handles branch coverage at the report level. So it is always collected into the internal data model, but whether it is shown in the report depends on the report type. We don't even implement `LCOV_EXCL_BR_LINE` markers (see #121).

The easiest solution will likely to write a filter that strips out the data from a gcvor-json report:

```bash
gcovr ... --json temp.json  # collect coverage
some_filter_to_strip_out_branch_coverage temp.json
gcovr -a temp.json ...  # create report
```

While I would like to see better exclusion mechanisms (also for branch coverage), I'm not sure whether it can reasonably done within gcovr. It is of course *technically* possible, but it feels to me like adding yet another special-purpose filtering mechanism isn't the best solution. Perhaps this rather points towards a need for scoped settings, e.g. that some configuration options should only be active in a particular directory. But that is currently infeasible because it would require a significant overhaul for config file management.

**Comment by Spacetown** [comment]
For me this will be a good feature. The filter is a easy way but the idea of a post processing phase between reading data and rendering as mentioned in https://github.com/gcovr/gcovr/issues/324#issuecomment-514960869 sounds realy good.
The first step could be moving the `GCOV_EXCL_START`/`GCOV_EXCL_STOP` into such a fix post processing and the user can add a python file with the option `--data-post-processing` which recieves the data and can set a ignore flag to true. At the end all ignored elements can be removed from the data model before rendering it.

**Comment by opikalo** [comment]
@latk I like yuour solution. Here is a gist if someone is trying to achieve something similar based on your suggestion.
<script src="https://gist.github.com/opikalo/1415210c0b97b18856193ab8944d7a70.js"></script>

**Comment by opikalo** [comment]
Workaround is good enough

**Comment by KantarBruceAdams** [comment]
I wasn't able to get the python solution to work but I was helped to create a solution using XSL to drop the branch coverage see https://stackoverflow.com/questions/71482426/xslt-remove-attribute-when-parent-node-has-an-attribute-matching-a-pattern-fil/71483170#71483170:

```
<xsl:stylesheet version="1.0" 
xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
<xsl:output method="xml" version="1.0" encoding="UTF-8" indent="yes"/>
<xsl:strip-space elements="*"/>

<!-- identity transform -->
<xsl:template match="@*|node()">
    <xsl:copy>
        <xsl:apply-templates select="@*|node()"/>
    </xsl:copy>
</xsl:template>

<xsl:template match="file[starts-with(@path, 'test/')]/lineToCover">
    <xsl:copy>
        <xsl:copy-of select="@lineNumber | @covered"/>
    </xsl:copy>
</xsl:template>

</xsl:stylesheet>
```

In this case the pattern is "test/". This could probably be improved to make that a parameter or perhaps read it from a file somehow.
